### PR TITLE
bazel: add execlog parser

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change

https://docs.bazel.build/versions/master/remote-execution-caching-debug.html advises to build the execlog parser to troubleshoot bazel cache issues. I would like to add it to our bazel packaging. So far fails with 

```
Analyzing: 2 targets (28 packages loaded, 323 targets configured)
    currently loading: src/main/protobuf
    Fetching @com_google_protobuf; fetching
WARNING: Download from https://mirror.bazel.build/github.com/protocolbuffers/protobuf/archive/v3.13.0.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Unknown host: mirror.bazel.build
Analyzing: 2 targets (28 packages loaded, 323 targets configured)
    currently loading: src/main/protobuf
WARNING: Download from https://github.com/protocolbuffers/protobuf/archive/v3.13.0.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Unknown host: github.com
Analyzing: 2 targets (28 packages loaded, 323 targets configured)
    currently loading: src/main/protobuf
ERROR: An error occurred during the fetch of repository 'com_google_protobuf':
   Traceback (most recent call last):
        File "/build/bazel_src/tools/build_defs/repo/http.bzl", line 111, column 45, in _http_archive_impl
                download_info = ctx.download_and_extract(
Error in download_and_extract: java.io.IOException: Error downloading [https://mirror.bazel.build/github.com/protocolbuffers/protobuf/archive/v3.13.0.tar.gz, https://github.com/protocolbuffers/protobuf/archive/v3.13.0.tar.gz] to /build/tmp.Yby7S1DIZl/.cache/bazel/_bazel_nixbld/96c321f6d4067432c019ff2725c18acb/external/com_google_protobuf/temp12648161410897702521/v3.13.0.tar.gz: Unknown host: github.com
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
You do not have network access in the build phase. Bazel has probably some phase where it does all the fetching.

**Comment by avdv** [comment]
The only thing needed is that you add the `com_google_protobuf` to the `srcDepsSet`, just like `rules_java` for example.

Then you only need to fix this error:
```
cp: cannot stat './bazel-src/bazel-bin/src/tools/execlog/parser': No such file or directory
```

**Comment by teto** [comment]
thanks that fixed the build process. On run:
```
/nix/store/54imygzblxfvirqdwy6pagdykaxjgy8b-bazel-3.7.2/bin/execlog -h                                                                                                                                         
/nix/store/54imygzblxfvirqdwy6pagdykaxjgy8b-bazel-3.7.2/bin/execlog: Cannot locate runfiles directory. (Set $JAVA_RUNFILES to inhibit searching.)
```
Seems like bin/bazel does a bit of magic upon start so not sure how to accomplish something similar or what's the best strategy there.

**Comment by avdv** [comment]
The execlog parser is a Java tool, packaged as a jar. The file you copied is just the startup script which tries to run `java -jar parser.jar`.

You need to copy the jar too and perhaps the script needs to be patched so that it can locate it properly.

**Comment by teto** [comment]
I've added a wrapper but reached the limit of my java skill : @avdv  Do you know how to solve this error when running execlog ?

```
Error: Could not find or load main class com.google.devtools.build.execlog.ExecLogParser
Caused by: java.lang.ClassNotFoundException: com.google.devtools.build.execlog.ExecLogParser
```

**Comment by avdv** [comment]
I would just add the `--singlejar` argument to the wrapper:
```
makeWrapper $out/bin/parser $out/bin/execlog --set JAVABIN '${openjdk}/bin/java' --set JAVA_RUNFILES $out/share/parser.runfiles --add-flags '--singlejar'
```
and I would not put those files into `out/bin`, but into `out/share` and just put the wrapper script into `bin`.

A further complication seems to be that the `parser` script wants the runfiles and these are local to the system. Does it run if you simply use `java -jar parser_deploy.jar` with a log file?

**Comment by avdv** [comment]
Yes, seems to work just fine with the fat jar:
```console
$ java -jar bazel-bin/src/tools/execlog/parser_deploy.jar --log_path ~/sandbox/haskell-bazel-nix-template/exec-1.log | head
command_args: "/nix/store/ijfv8z733l3kwzigvvilc21hpyqqvvd5-bash/bin/bash"
command_args: "-c"
command_args: "source external/bazel_tools/tools/genrule/genrule-setup.sh; cp xmobar/paths.in bazel-out/k8-fastbuild/bin/xmobar/Paths_xmobar.hs"
environment_variables {
  name: "PATH"
  value: "/home/claudio/.rbenv/shims:/home/claudio/.rbenv/bin:/home/claudio/.jenv/shims:/home/claudio/.rbenv/shims:/home/claudio/.rbenv/bin:/home/claudio/.local/bin:/home/claudio/bin:/home/claudio/.local/bin:/run/wrappers/bin:/home/claudio/.nix-profile/bin:/etc/profiles/per-user/claudio/bin:/nix/var/nix/profiles/default/bin:/run/current-system/sw/bin:/home/claudio/bin:/home/claudio/.local/bin:/home/claudio/go/bin:/home/claudio/.gem/ruby/2.6.0/bin:/home/claudio/.npm-packages/bin:/home/claudio/.jenv/bin:/home/claudio/.githubcli/bin:/nix/store/2jysm3dfsgby5sw5jgj43qjrb5v79ms9-bash-4.4-p23/bin:/nix/store/w9wc0d31p4z93cbgxijws03j5s2c4gyf-coreutils-8.31/bin:/nix/store/aja0dimyn0sg5b9zf1cav4k43p8h5xqc-findutils-4.7.0/bin:/nix/store/4z79ipgxqn80ns7mpax25zmb77i3ndfw-gawk-5.1.0/bin:/nix/store/xhvk95cjr2dk339airqxqfk04c0zras6-gnugrep-3.4/bin:/nix/store/1yyn6aar4kw3vjn2fs5xv5jxywdly8wn-gnutar-1.32/bin:/nix/store/mgkwf0c30x2xdw98nh7kj40n83wwv9sf-gnused-4.8/bin:/nix/store/ljjm1r3hn8wmlh5gp38vms13hipjmygy-gzip-1.10/bin:/nix/store/zwms4mg5cf5vmzg9wlskl54fck3643z4-which-2.21/bin:/nix/store/bs0d5iz7isfa8spbp2d4w31sbkkbd6ff-unzip-6.0/bin:/nix/store/y6kf7b0bf91if7i11pk84rxszvbggnqk-file-5.39/bin:/nix/store/jcbs5xa84wk053c7mnhhl2998j97vslw-zip-3.0/bin"
...
```

**Comment by teto** [comment]
thanks. should I ditch the wrapper and call the fatjar directly then ?

**Comment by avdv** [comment]
> thanks. should I ditch the wrapper and call the fatjar directly then ?

Yes, only copy the jar and create a simple script in `bin` calling it with all arguments passed through.

Maybe we should call the script something other than `parser` though. How about `bazel-execlog-parser`?!

**Comment by teto** [comment]
it should work now ! should I put it in bazel_4 as well ?

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix [hunk 1]
```diff
@@ -62,6 +64,7 @@ let
       srcs.rules_cc
       srcs.rules_java
       srcs.rules_proto
+      srcs.com_google_protobuf
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:67]: We could use ``with srcs;`` here.

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix [hunk 2]
```diff
@@ -527,6 +530,15 @@ stdenv.mkDerivation rec {
         --output=./bazel_src/output/bazel-complete.bash \
         --prepend=./bazel_src/scripts/bazel-complete-header.bash \
         --prepend=./bazel_src/scripts/bazel-complete-template.bash
+
+    # need to change directory for bazel to find the workspace
+    cd ./bazel_src
+    # build execlog tooling
+    export HOME=$(mktemp -d)
+    # --output_base=./cache
```
Comment (SuperSandro2000) [inline pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:538]: Is this from testing?

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix [hunk 3]
```diff
@@ -22,6 +22,8 @@
 , file
 , substituteAll
 , writeTextFile
+# for execlog binary
+, openjdk
```
Comment (avdv) [inline pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:26]: no need to introduce a third JDK, we already have `buildJdk` and `runJdk`, just use the latter.

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix (hunk 2 above)
Comment (teto) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:538]: indeed my bad

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix (hunk 3 above)
Comment (teto) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix]: done

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix (hunk 1 above)
Comment (teto) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:65]: `with srcs;` is not a net improvement IMO so I'll let it be.

**Comment by teto** [comment]
I've tried to address reviews.

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix [hunk 4]
```diff
@@ -537,7 +546,15 @@ stdenv.mkDerivation rec {
     # The binary _must_ exist with this naming if your project contains a .bazelversion
     # file.
     cp ./bazel_src/scripts/packages/bazel.sh $out/bin/bazel
+
+    mkdir $out/share
+    cp ./bazel_src/bazel-bin/src/tools/execlog/parser_deploy.jar $out/share/parser_deploy.jar
     mv ./bazel_src/output/bazel $out/bin/bazel-${version}-${system}-${arch}
+    cat <<EOF > $out/bin/execlog
```
Comment (avdv) [inline pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:553]: I am not really happy with the name. If you install bazel in your profile, you will get a `execlog` program in your PATH... I'd really prefer `bazel-execlog` or something.

Comment (teto) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix]: Valid point, I've renamed it to bazel-execlog

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:65]: If we you add it in line 45 we can remove all those srcs.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 111975` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
10 packages marked as broken and skipped:
</summary>
<ul>
<li>gvisor</li>
<li>python38Packages.arviz</li>
<li>python38Packages.dm-sonnet</li>
<li>python38Packages.graph_nets</li>
<li>python38Packages.optuna</li>
<li>python38Packages.pymc3</li>
<li>python38Packages.rl-coach</li>
<li>python39Packages.dm-sonnet</li>
<li>python39Packages.graph_nets</li>
<li>python39Packages.rl-coach</li>
</ul>
</details>
<details>

  <summary>4 packages failed to build and are new build failures:</summary>
  <ul>
<li>bazel: <a href="https://termbin.com/t9n7">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/5fff5265cc5583c4a0002945378976e5/raw/index.html">pretty log </a></li>
<li>(bazel_3): log was empty</li>
<li>bazel-remote: log was empty</li>
<li>bazel-watcher: log was empty</li>
  </ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
bazel:
<details>


warning: attribute-typo
A likely typo in `preBuildPhase` argument was found, did you mean `preBuildPhases` or `buildPhase`?

Near pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:516:3:

```
    |
516 |   preBuildPhase = ''
    |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/attribute-typo.md
</details>
</details>

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix [hunk 5]
```diff
@@ -550,7 +550,7 @@ stdenv.mkDerivation rec {
     mkdir $out/share
     cp ./bazel_src/bazel-bin/src/tools/execlog/parser_deploy.jar $out/share/parser_deploy.jar
     mv ./bazel_src/output/bazel $out/bin/bazel-${version}-${system}-${arch}
-    cat <<EOF > $out/bin/execlog
+    cat <<EOF > $out/bin/bazel-execlog
     #!${runtimeShell} -e
     ${runJdk}/bin/java -jar $out/share/parser_deploy.jar \$@
     EOF
```
Comment (avdv) [inline pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:556]: you also have to change the name in the chmod call

Comment (teto) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:556]: I am ashamed ! sorry :'( fixed !

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix [hunk 6]
```diff
@@ -527,6 +528,14 @@ stdenv.mkDerivation rec {
         --output=./bazel_src/output/bazel-complete.bash \
         --prepend=./bazel_src/scripts/bazel-complete-header.bash \
         --prepend=./bazel_src/scripts/bazel-complete-template.bash
+
+    # need to change directory for bazel to find the workspace
+    cd ./bazel_src
+    # build execlog tooling
+    export HOME=$(mktemp -d)
+    ./output/bazel build  src/tools/execlog:all
+    ./output/bazel build  src/tools/execlog:parser_deploy.jar
```
Comment (avdv) [inline pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:537]: Technically, you only need to build the `src/tools/execlog:parser_deploy.jar` target. But should not hurt if you keep the first command.

**Review by avdv — APPROVED** [review]
LGTM

**Comment by avdv** [comment]
> it should work now ! should I put it in bazel_4 as well ?

Yes, this would be nice; once we got this sorted out.

File: pkgs/development/tools/build-managers/bazel/bazel_3/default.nix (hunk 6 above)
Comment (teto) [reply pkgs/development/tools/build-managers/bazel/bazel_3/default.nix:536]: indeed, I thought about removing it but it's such a small build. Your comment lead me to remove it.

**Comment by teto** [comment]
you mean in a separate commit ? actually I might prefer that to speed up the arrival of this in the cache.

**Comment by avdv** [comment]
> you mean in a separate commit ? actually I might prefer that to speed up the arrival of this in the cache.

Separate commit, yes. But you could do it in this PR I guess. But I have no idea how / when something hits the cache...

**Comment by teto** [comment]
I've added similar instructions for bazl4. Tested locally and it looked fine.

**Comment by SuperSandro2000** [comment]
> But I have no idea how / when something hits the cache...

When the commit is in a hydra eval set for master it should hit the cache of it builds and had not set hydra platforms to an empty list.

**Comment by teto** [comment]
@avdv  I deem it ready. Could you merge if it's ok for you ?

**Comment by avdv** [comment]
> @avdv I deem it ready. Could you merge if it's ok for you ?

I cannot, I have no merging permissions.

**Comment by teto** [comment]
haha I could've waited for long then. Thanks for the reviews. Going to merge then :)

**Pull request merged**